bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java

```diff
@@ -298,7 +303,27 @@ public final void run() {
 			if (getOperation() instanceof IAdvancedUndoableOperation2) {
 				runInBackground = ((IAdvancedUndoableOperation2) getOperation()).runInBackground();
 			}
-			progressDialog.run(runInBackground, true, runnable);
+			if (runInBackground) {
+				progressDialog.run(runInBackground, true, runnable);
+			} else {
+				// prevent UI freeze during AutoBuild as in 
+				// org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper#perform
+				final IJobManager manager = Job.getJobManager();
+				// pessimistic scheduling rule:
+				final ISchedulingRule rule = ResourcesPlugin.getWorkspace().getRoot(); 
+				try {
+					try {
+						// interrupt autobuild or show Wait/Cancel Dialog:
+						Runnable r = () -> manager.beginRule(rule, null);
```

Comment:
This change causes a regression. If the Undo task joins on a workspace rule, it hangs now.
So starting a rule here is not OK.
I would propose to revert the change.

Reply:
> If the Undo task joins on a workspace rule

how to reproduce that?

Reply:
Have custom `IUndoableOperation` (we have one that extends `AbstractOperation`) that runs operation in a different thread and there takes the workspace rule.

```
    @Override
    public IStatus undo(IProgressMonitor monitor, IAdaptable info) throws ExecutionException {

        IProgressService progressService = PlatformUI.getWorkbench().getProgressService();
        try {
            progressService.run(true, false, new IRunnableWithProgress() {
                @Override
                public void run(IProgressMonitor monitor) throws InvocationTargetException, InterruptedException {
                    ISchedulingRule rule = ResourcesPlugin.getWorkspace().getRoot();
                   Job.getJobManager().beginRule(rule, monitor); // here we hang now
                   // ... do something that needs workspace
                }
            });
        } catch (InvocationTargetException | InterruptedException e) {
        }

        return Status.OK_STATUS;
    }
```

Reply:
Basically, locking **anything accessible to 3rd party code** before starting undo is not OK, because "undo" in 3rd party code can require the same lock.

Also interesting (I haven't noticed that first), that the org.eclipse.ui.workbench bundle in question had no references to org.eclipse.core.resources bundle before, I think for a good reason.